filters/include/pcl/filters/impl/farthest_point_sampling.hpp

```diff
@@ -37,71 +35,92 @@
  *
  */
 
-#ifndef PCL_FILTERS_FARTHEST_POINT_SAMPLING_IMPL_H_
-#define PCL_FILTERS_FARTHEST_POINT_SAMPLING_IMPL_H_
+#pragma once
 
-#include <limits>
-#include <algorithm>
-#include <pcl/filters/farthest_point_sampling.h>
 #include <pcl/common/geometry.h>
+#include <pcl/filters/farthest_point_sampling.h>
 #include <pcl/point_cloud.h>
 #include <pcl/point_types.h>
-
+#include <algorithm>
+#include <limits>
+#include <random>
 
 template<typename PointT> void
 pcl::FarthestPointSampling<PointT>::applyFilter (PointCloud &output)
 {
-    int initial_size = input_->points.size();
-    int sample_size = sample_;
-    int max_index;
-    std::vector<int> selected_indices;
-    std::vector<float>distances_to_selected_points;
-    
-    //set random seed
-    std::srand (seed_);
+  std::vector<int> indices;
+  if (keep_organized_)
+  {
+    bool temp = extract_removed_indices_;
+    extract_removed_indices_ = true;
+    applyFilter (indices);
+    extract_removed_indices_ = temp;
 
-    
-    //check if number of points in cloud is less than requested number of points
-    if (initial_size <= sample_size)
+    output = *input_;
+    for (const auto ri : *removed_indices_)  // ri = removed index
+      output.points[ri].x = output.points[ri].y = output.points[ri].z = user_filter_value_;
+    if (!std::isfinite (user_filter_value_))
+      output.is_dense = false;
+  }
+  else
+  {
+    output.is_dense = true;
+    applyFilter (indices);
+    pcl::copyPointCloud (*input_, indices, output);
+  }
+}
+
+template<typename PointT> void
+pcl::FarthestPointSampling<PointT>::applyFilter (std::vector<int> &indices)
+{
+  std::size_t size = input_->points.size();
+  int max_index;
+  std::vector<float> distances_to_selected_points (size, std::numeric_limits<float>::max ());
+  
+  //set random seed
+  std::mt19937 random_gen(seed_);
+  std::uniform_int_distribution<> dis(0, size -1);
+  
+  //check if number of points in cloud is less than requested number of points
+  if (size <= sample_)
+  {
+    PCL_THROW_EXCEPTION (BadArgumentException,
+                        "Requested number of points is greater than point cloud size!");
+    return;
+  }
+
+  //pick the first point at random
+  max_index = dis(random_gen);
+  distances_to_selected_points[max_index] = -1.0;
+  indices.push_back(max_index);
+
+  for (int j = 0; j < sample_ - 1; j++)
+  {
+    //recompute distances
+    for (int i = 0; i < distances_to_selected_points.size(); i++)
     {
-        PCL_THROW_EXCEPTION (BadArgumentException,
-                           "Requested number of points is greater than point cloud size!");
-        return;
+      if (distances_to_selected_points[i] == -1.0)
```

Comment:
>If it's assured that the distance will always be more than 0

since the distance calculation is based on Eigen's norm function I think it's fair to say that it wil be greater than or equal to zero. 
>we can simplify this loop using valid_distance_to_current_pt name instead of distances_to_selected_point

I don't quite get what you mean here though. Do you mean I should remove irrelevant points from this vector altogether?

Reply:
1. If the distance is strictly more than zero (ie, all points are unique)
2. Then we can use zero as sentinel instead of -1
3. Which means the check can be `if (value[i] != 0)`
4. Which can be expressed using a prefix `valid_`
5. This allows the check to be written as `if (valid_value[i])`

Reply:
Oh right, thanks for the explanation :)  Yes that does indeed make sense. But I'm a bit worried about the possible existence of duplicated points? I feel like that might be somewhat common, especially in clouds where multiple scans are consolidated to one cloud? In such a case the distance would be zero. So would it make more sense to use -1?

Reply:
But we don't want to visit a point multiple times in such a sampling.

Reply:
Duplicated points are apparently an issue even in R (which is supposed to be nice for statistical work). See [the official documentation](https://rdrr.io/cran/rdist/man/farthest_point_sampling.html). With the following code:
```r
library(rdist)

# generate data
df <- matrix(c(0,0,1,1,0,0,1,1,0,0,1,1), ncol = 2)
dist_mat <- pdist(df)
# farthest point sampling
fps <- farthest_point_sampling(dist_mat)
fps2 <- farthest_point_sampling(df, metric = "euclidean")
all.equal(fps, fps2)
# have a look at the fps distance matrix
# rdist(df[fps[1:5], ])
dist_mat[fps, fps][1:5, 1:5]
fps
```
we get an unexpected output of `[1,3,2,4,5,6]` instead of `[1,2,3,4,5,6]`.

Using 0 will reduce the size and not keep it constant if that's what is requested by the user. -1 seems to be better if there are non-unique points.

Reply:
Weird that R would have such a flaw. Anyway I suppose I better keep it at -1 for now.